File: YuXiangROS/Chap5/5.2_5.3_TF_ws/src/demo_cpp_tf/src/static_tf_broadcaster.cpp
#include <geometry_msgs/msg/transform_stamped.hpp> //提供消息接口
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <tf2/LinearMath/Quaternion.h>             // 提供四元数接口
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp> // 提供消息类型转换函数
#include <tf2_ros/static_transform_broadcaster.h>  // 提供静态坐标变换广播类

class StaticTFBroadcaster : public rclcpp::Node {
  private:
    // 静态坐标变换广播智能指针
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_broadcaster_;

  public:
    StaticTFBroadcaster() : Node("static_tf_broadcaster") {
        // 创建静态坐标变换广播器
        static_broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        this->publish_static_tf_();
    }

  private:
    void publish_static_tf_() {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "target_point";

        transform.transform.translation.x = 5.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;

        tf2::Quaternion quat;
        quat.setRPY(0, 0, 60 * M_PI / 180); // 弧度制欧拉角转四元数
        transform.transform.rotation =
            tf2::toMsg(quat); // 直接转成消息接口类型 (也可以仿照Python写法)

        static_broadcaster_->sendTransform(transform);
    };
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: YuXiangROS/Chap5/5.2_5.3_TF_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
"""
1. 命令行发布静态坐标变换关系
ros2 run tf2_ros static_transform_publisher --x 0.1 --y 0.2 --z 0.3 --roll 0.0 --pitch 0.0 --yaw 0.0 --frame-id base_link --child-frame-id base_laser
2. 使用命令行计算坐标系变换关系
ros2 run tf2_ros tf2_echo base_link wall_point
3. 查看坐标系连接关系(生成 PDF 和 GV)
ros2 run tf_tool view_frames
"""

# 角度与弧度转换
import math

import rclpy
from geometry_msgs.msg import TransformStamped
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster
from tf_transformations import quaternion_from_euler


class StaticTFBroadcaster(Node):
    def __init__(self):
        super().__init__("static_tf2_broadcaster")

        # 创建静态坐标转换广播器
        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        self.publish_static_tf()

    def publish_static_tf(self):
        # 创建静态坐标转换消息
        transform = TransformStamped()

        # 获取当前时间并转换为消息时间戳
        transform.header.stamp = self.get_clock().now().to_msg()
        # 父坐标系名称
        transform.header.frame_id = "base_link"
        # 子坐标系名称
        transform.child_frame_id = "camera_link"

        # 空间平移
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6

        # 空间旋转
        # 旋转180度，绕x轴; 转换为四元数
        rotation_quat = quaternion_from_euler(math.radians(180), 0, 0)
        transform.transform.rotation.x = rotation_quat[0]
        transform.transform.rotation.y = rotation_quat[1]
        transform.transform.rotation.z = rotation_quat[2]
        transform.transform.rotation.w = rotation_quat[3]

        # 发布静态坐标转换消息
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f"Publish TF: {transform}")


def main(args=None):
    rclpy.init(args=args)
    node = StaticTFBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
